tfhe/src/core_crypto/algorithms/lwe_multi_bit_bootstrap_key_generation.rs

```diff
@@ -450,3 +451,324 @@ where
+
+    generate_seeded_lwe_multi_bit_bootstrap_key(
+        input_lwe_secret_key,
+        output_glwe_secret_key,
+        &mut bsk,
+        noise_parameters,
+        noise_seeder,
+    );
+
+    bsk
+}
+
+/// Parallel variant of [`generate_seeded_lwe_multi_bit_bootstrap_key`], it is recommended to use
+/// this function for better key generation times as LWE bootstrapping keys can be quite large.
+#[allow(clippy::too_many_arguments)]
+pub fn par_generate_seeded_lwe_multi_bit_bootstrap_key<
+    Scalar,
+    InputKeyCont,
+    OutputKeyCont,
+    OutputCont,
+    NoiseSeeder,
+>(
+    input_lwe_secret_key: &LweSecretKey<InputKeyCont>,
+    output_glwe_secret_key: &GlweSecretKey<OutputKeyCont>,
+    output: &mut SeededLweMultiBitBootstrapKey<OutputCont>,
+    noise_parameters: impl DispersionParameter + Sync,
+    noise_seeder: &mut NoiseSeeder,
+) where
+    Scalar: UnsignedTorus + CastFrom<usize> + Sync + Send,
+    InputKeyCont: Container<Element = Scalar>,
+    OutputKeyCont: Container<Element = Scalar> + Sync,
+    OutputCont: ContainerMut<Element = Scalar>,
+    // Maybe Sized allows to pass Box<dyn Seeder>.
+    NoiseSeeder: Seeder + ?Sized,
+{
+    assert!(
+        output.input_lwe_dimension() == input_lwe_secret_key.lwe_dimension(),
+        "Mismatched LweDimension between input LWE secret key and LWE bootstrap key. \
+        Input LWE secret key LweDimension: {:?}, LWE bootstrap key input LweDimension {:?}.",
+        input_lwe_secret_key.lwe_dimension(),
+        output.input_lwe_dimension()
+    );
+
+    assert!(
+        output.glwe_size() == output_glwe_secret_key.glwe_dimension().to_glwe_size(),
+        "Mismatched GlweSize between output GLWE secret key and LWE bootstrap key. \
+        Output GLWE secret key GlweSize: {:?}, LWE bootstrap key GlweSize {:?}.",
+        output_glwe_secret_key.glwe_dimension().to_glwe_size(),
+        output.glwe_size()
+    );
+
+    assert!(
+        output.polynomial_size() == output_glwe_secret_key.polynomial_size(),
+        "Mismatched PolynomialSize between output GLWE secret key and LWE bootstrap key. \
+        Output GLWE secret key PolynomialSize: {:?}, LWE bootstrap key PolynomialSize {:?}.",
+        output_glwe_secret_key.polynomial_size(),
+        output.polynomial_size()
+    );
+
+    let mut generator = EncryptionRandomGenerator::<ActivatedRandomGenerator>::new(
+        output.compression_seed().seed,
+        noise_seeder,
+    );
+
+    let gen_iter = generator
+        .par_fork_multi_bit_bsk_to_ggsw_group::<Scalar>(
+            output.input_lwe_dimension(),
+            output.decomposition_level_count(),
+            output.glwe_size(),
+            output.polynomial_size(),
+            output.grouping_factor(),
+        )
+        .unwrap();
+
+    let grouping_factor = output.grouping_factor();
+    let ggsw_per_multi_bit_element = grouping_factor.ggsw_per_multi_bit_element();
+
+    output
+        .par_iter_mut()
+        .chunks(ggsw_per_multi_bit_element.0)
```

Comment:
probable, I'll check

Reply:
looks like I did not write that par_chunk_mut function I would propose to add an internal issue to correct it rather than correct here ? Otherwise it will delay the merge of this PR and I'm going to be away

Reply:
https://github.com/zama-ai/tfhe-rs-internal/issues/128